datasets/flwr_datasets/partitioner/grouped_natural_id_partitioner.py

```diff
@@ -0,0 +1,219 @@
+        The number of unique ids that will be placed in a single group.
+    mode: Literal["allow-smaller", "allow-bigger", "drop-reminder", ""strict"]
+        The mode that will be used to handle the remainder of the unique ids.
+        - "allow-smaller": The last group can be smaller than the group_size.
+        - "allow-bigger": The first group can be bigger than the group_size.
+        - "drop-reminder": The last group will be dropped if it is smaller than the
+        group_size.
+        - "strict": Raises a ValueError if the remainder is not zero. In this mode, you
+        expect each group to have the same size.
+    sort_unique_ids: bool
+        If True, the unique natural ids will be sorted before creating the groups.
+
+    Examples
+    --------
+    Partition users in the "sentiment140" (aka Twitter) dataset into groups of two
+    users following the default mode:
+
+    >>> from flwr_datasets import FederatedDataset
+    >>> from flwr_datasets.partitioner import GroupedNaturalIdPartitioner
+    >>>
+    >>> partitioner = GroupedNaturalIdPartitioner(partition_by="user", group_size=2)
+    >>> fds = FederatedDataset(dataset="sentiment140",
+    >>>                        partitioners={"train": partitioner})
+    >>> partition = fds.load_partition(0)
+    """
+
+    def __init__(
+        self,
+        partition_by: str,
+        group_size: int,
+        mode: Literal[
+            "allow-smaller", "allow-bigger", "drop-reminder", "strict"
+        ] = "allow-smaller",
+        sort_unique_ids: bool = False,
+    ) -> None:
+        super().__init__()
+        self._partition_id_to_natural_ids: Dict[int, List[Any]] = {}
+        self._natural_id_to_partition_id: Dict[Any, int] = {}
+        self._partition_id_to_indices: Dict[int, NDArrayInt] = {}
+        self._partition_by = partition_by
+        self._mode = mode
+        self._sort_unique_ids = sort_unique_ids
+
+        if group_size < 0:
+            raise ValueError("group_size must be a positive integer")
+        self._group_size = group_size
+
+    def _create_int_partition_id_to_natural_id(self) -> None:
+        """Create a mapping from int indices to unique client ids from dataset.
+
+        Natural ids come from the column specified in `partition_by`.
+        """
+        unique_natural_ids = self.dataset.unique(self._partition_by)
+        if self._mode != "allow-smaller" and self._group_size > len(unique_natural_ids):
+            raise ValueError(
+                "The group size needs to be smaller than the number of the unique "
+                "natural ids unless you are using allow-smaller mode which will "
+                "result in a single partition."
+            )
+        if self._sort_unique_ids:
+            unique_natural_ids = sorted(unique_natural_ids)
+        num_unique_natural_ids = len(unique_natural_ids)
+        remainder = num_unique_natural_ids % self._group_size
+        num_groups = num_unique_natural_ids // self._group_size
```

Comment:
When `group_size` is bigger than `num_unique_natural_ids` (something allowed for `allow-smaller`) the resulting  `num_groups = 0`. This will make `np.array_split(...)` call raise an error.

Reply:
I wonder if we should do instead

```python
num_groups = num_unique_natural_ids // self._group_size
```
to this:
```
num_groups = max(1, num_unique_natural_ids // self._group_size)
```

Reply:
Thanks for this catch. I applied a bit differently because this would produce an overlapping sample; I additionally zeroed the remainder. Why? Not to trigger the additional appends based on the reminder later in the code.
```
        if num_groups == 0 and self._mode == "allow-smaller":
            num_groups = 1
            remainder = 0
```
I've added a new test to check this scenario and extended the overlap test to catch something like that.